exporting/aws_kinesis/aws_kinesis.c

```diff
@@ -0,0 +1,157 @@
+    struct aws_kinesis_specific_data *connector_specific_data = callocz(1, sizeof(struct aws_kinesis_specific_data));
+    instance->connector_specific_data = (void *)connector_specific_data;
+
+    kinesis_init(
+        (void *)connector_specific_data,
+        instance->config.destination,
+        connector_specific_config->auth_key_id,
+        connector_specific_config->secure_key,
+        instance->config.timeoutms);
+
+    return 0;
+}
+
+/**
+ * AWS Kinesis connector worker
+ *
+ * Runs in a separate thread for every instance.
+ *
+ * @param instance_p an instance data structure.
+ */
+void aws_kinesis_connector_worker(void *instance_p)
+{
+    struct instance *instance = (struct instance *)instance_p;
+    struct aws_kinesis_specific_config *connector_specific_config = instance->config.connector_specific_config;
+    struct aws_kinesis_specific_data *connector_specific_data = instance->connector_specific_data;
+
+    while (!netdata_exit) {
+        unsigned long long partition_key_seq = 0;
+        struct stats *stats = &instance->stats;
+
+        uv_mutex_lock(&instance->mutex);
+        uv_cond_wait(&instance->cond_var, &instance->mutex);
+
+        BUFFER *buffer = (BUFFER *)instance->buffer;
+        size_t buffer_len = buffer_strlen(buffer);
+
+        size_t sent = 0;
+
+        while (sent < buffer_len) {
+            char partition_key[KINESIS_PARTITION_KEY_MAX + 1];
+            snprintf(partition_key, KINESIS_PARTITION_KEY_MAX, "netdata_%llu", partition_key_seq++);
+            size_t partition_key_len = strnlen(partition_key, KINESIS_PARTITION_KEY_MAX);
+
+            const char *first_char = buffer_tostring(buffer) + sent;
+
+            size_t record_len = 0;
+
+            // split buffer into chunks of maximum allowed size
+            if (buffer_len - sent < KINESIS_RECORD_MAX - partition_key_len) {
+                record_len = buffer_len - sent;
+            } else {
+                record_len = KINESIS_RECORD_MAX - partition_key_len;
+                while (*(first_char + record_len) != '\n' && record_len)
+                    record_len--;
+            }
+            char error_message[ERROR_LINE_MAX + 1] = "";
+
+            debug(
+                D_BACKEND,
+                "EXPORTING: kinesis_put_record(): dest = %s, id = %s, key = %s, stream = %s, partition_key = %s, \
+                  buffer = %zu, record = %zu",
+                instance->config.destination,
+                connector_specific_config->auth_key_id,
+                connector_specific_config->secure_key,
+                connector_specific_config->stream_name,
+                partition_key,
+                buffer_len,
+                record_len);
+
+            kinesis_put_record(
+                connector_specific_data, connector_specific_config->stream_name, partition_key, first_char, record_len);
+
+            sent += record_len;
+            stats->chart_transmission_successes++;
```

Comment:
Do you want to count this success if the `kinesis_get_result` indicates that there was a failure?

Reply:
No. It will be used for displaying internal statistics which is not implemented yet in the exporting engine.